ml-prims/src/selection/columnWiseSort.h

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include <cstddef>
+#include "cuda_utils.h"
+
+#include <cub/cub.cuh>
+#include <limits>
+
+#define ALIGN_BYTE 256
+#define ALIGN_MEMORY(x) (x + ALIGN_BYTE - 1) & ~(ALIGN_BYTE - 1)
+
+#define VALIDATE_TEMPLATE (is_same< InType, float >::value && BLOCK_SIZE <= 512) || \
```

Comment:
It's probably better to use a templated struct (with an enum or typedef inside it which does this check), right?

Reply:
Can you give a sample code here, I don't understand how you mean to do it.

Reply:
My bad. I should have done that before. Here's what I meant:
```cpp
template <typename InType, typename OutType>
struct Checker {
  enum {
    IsValid = (std::is_same< InType, float >::value && BLOCK_SIZE <= 512) ||
                  (std::is_same< InType, double >::value &&
                   (std::is_same< OutType, int >::value || std::is_same< OutType, float >::value) &&
                    BLOCK_SIZE <= 512 && ITEMS_PER_THREAD <= 8) ||
                 (std::is_same< InType, double >::value && std::is_same< OutType, double >::value &&
                  BLOCK_SIZE <= 512 && ITEMS_PER_THREAD <= 6)
  };
};
```
Then, all you got to do is to use it as: `Checker<InType, OutType>::IsValid`. Hope this helps.